segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+                return null;
+            });
+        }
+        return result;
+    }
+
+    /**
+     * Appends the result of the given serialization function conditionally based on the current length of the Attribute Segment.
+     * This method will retry the write subject to the APPEND_RETRY policy.
+     *
+     * @param getSerialization A Supplier that, when invoked, returns a CompletableFuture whose result will be the requested
+     *                         serialization.
+     * @param timer            Timer for the operation. Used for timeouts.
+     * @return A CompletableFuture that, when completed, will contain information about the data that was just written.
+     */
+    private CompletableFuture<WriteInfo> appendConditionally(Supplier<CompletableFuture<ArrayView>> getSerialization, TimeoutTimer timer) {
+        return APPEND_RETRY.runAsync(() -> appendConditionallyOnce(getSerialization, timer), this.executor);
+    }
+
+    /**
+     * Appends the result of the given serialization function conditionally based on the current length of the Attribute Segment.
+     * This method does not perform any retries.
+     *
+     * @param getSerialization A Supplier that, when invoked, returns a CompletableFuture whose result will be the requested
+     *                         serialization.
+     * @param timer            Timer for the operation. Used for timeouts.
+     * @return A CompletableFuture that, when completed, will contain information about the data that was just written.
+     */
+    private CompletableFuture<WriteInfo> appendConditionallyOnce(Supplier<CompletableFuture<ArrayView>> getSerialization, TimeoutTimer timer) {
+        // We want to make sure that the serialization we generate is accurate based on the state of the Attribute Segment.
+        // This is to protect against potential corruptions due to concurrency: for example we picked data for a Snapshot,
+        // then merged it with some other changes, then wrote it back - we want to ensure nobody else wrote anything in the meantime.
+        // As such, we need to do a conditional append keyed on the length of the Attribute Segment. Should there be
+        // a concurrent change, we will need to re-generate the serialization in order to guarantee that we always write
+        // the latest data.
+        AttributeSegment as = this.attributeSegment.get();
+        if (isMainSegmentDeleted()) {
+            // Verify if the main segment has been deleted and stop now if needed.
+            return handleMainSegmentDeleted(timer.getRemaining());
+        }
+
+        long offset = as.getLength();
+        return getSerialization.get().thenComposeAsync(data ->
+                Futures.exceptionallyCompose(
+                        this.storage
+                                .write(as.handle, offset, data.getReader(), data.getLength(), timer.getRemaining())
```

Comment:
I'm interested in the recovery path here. From the code, `write` throws `BadOffsetException` when offset doesn't match length. In the presence of a concurrent write, the offset check will fail and it will throw. I suppose that the snapshot in the attribute segment is consistent because of the semantics of conditional append. 

Retrying seems fine because a successful append is conditional on the instance knowing the previous snapshot so that the merge is consistent. The only possibility of a problem that I'm wondering about is the one of the number of retries exhausting. In this case, would the segment container stop?

Reply:
No, we only stop the container in case of `DataCorruptionException`, which can only be thrown upon reading (if we find a corrupted AttributeIndex).

If we get two or more concurrent calls, one of them will succeed and the rest will fail. Your scenario will happen if we have more than 10 concurrent calls (max retries == 10) and two of these will be so unlucky that they will execute at the exact same time (with an exp backoff, the probability of this is pretty remote). In that case, they will simply fail, and the `StorageWriter` will just reattempt them later. 

It may be worthwhile to say that this concurrent append thing was added for future-proofing. Currently, there is only one append to any Segment Attribute Index going on at any given time, due to the sequential nature of the Storage Writer.